### PR TITLE
doc: add systemd chapter

### PR DESCRIPTION
**Pull request by happysalada**
###### Motivation for this change

Systemd has many features, some useful to create modules, but most features are hard to discover and use if you don't already know about them. This is an attempt at covering the basics that are useful in nixpkgs
see https://github.com/NixOS/nixpkgs/issues/102397 for more details.

Please note that this is still an overview, I would like feedback on 
- Potential secctions that you think are important but you feel are missing
- Better order of the section to create a more natural flow.
- Better place to put that documentation ?

I will populate each sections shortly, ideally there is consensus on the sections and their order before that.

###### Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/#sec-conf-file))
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes (or backporting 21.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by happysalada** [comment]
It works like a charm and I can't see anything I would change in the code.
I'm happy to merge this. Are you looking for more feedback ?

**Comment by jtojnar** [comment]
Should not this rather go to NixOS manual?

**Comment by happysalada** [comment]
This should indeed go to nixos manual.
let me also have a look at merging existing docs on it.
ryan also mentioned adding a chapter on tmpfiles on matrix.

File: nixos/doc/manual/development/systemd-environment-and-secrets.section.md [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+Environment variables should not contain secret data.
+The main reason being that environment variables are accessible to unpriviledged users.
+See [systemd doc](https://www.freedesktop.org/software/systemd/man/systemd.exec.html#Environment=) for a more detailed explanation.
+
+### Secret management {#sec-systemd-secrets}
+
+To pass secret data to a systemd service, systemd has the `LoadCredential` mechanism.
+By passing a file path to systemd, it will make sure that the file is available to the process under a specific path.
+The load credential attribute takes an id a colon and a file path
+let's have a look at an [example](https://github.com/NixOS/nixpkgs/blob/master/nixos/modules/services/web-apps/lemmy.nix#L182)
```
Comment (jian-lin) [inline nixos/doc/manual/development/systemd-environment-and-secrets.section.md:21]: Maybe a permalink like this `https://github.com/NixOS/nixpkgs/blob/3b0bff383f8e894b1ee7a30b62c8b727a68511df/nixos/modules/services/web-apps/lemmy.nix#L182` is better?

Comment (happysalada) [reply nixos/doc/manual/development/systemd-environment-and-secrets.section.md:21]: Thinking about those, I thought it would be better to just include the code-snippet rather than a link.
My next iteration will just have the relevant snippets just copied.

File: nixos/doc/manual/development/systemd-environment-and-secrets.section.md [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+```
+LoadCredential = "jwt_secret:${cfg.jwtSecretPath}";
+```
+Inside your application, the jwt secret will be available under `$CREDENTIALS_DIRECTORY/jwt_secret` environment variable.
+
+Note that this requires your application to read secrets from a file.
+
+If your service can only read secrets through environment variables, the best alternative is to ask the end user to pass an `environmentFile` containing the secrets.
```
Comment (domenkozar) [inline nixos/doc/manual/development/systemd-environment-and-secrets.section.md:29]: Would be nice to document how that's done or link to an example

**Review by kevincox — APPROVED** [review]

**Comment by happysalada** [comment]
Thanks a lot for commenting to both of you!

I have this task for the holidays to finish this PR.
- I need to replace the links with the actual code block 
- I need to add 2 examples of 
  - How to pass an environment file
  - How to use genJqSecretsReplacementSnippet (an alternative to replace a secret. There is an example in es7 (in a PR currently https://github.com/NixOS/nixpkgs/pull/150879).

File: nixos/doc/manual/development/systemd-state-management.section.md [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+State refers to files and directories for config or running a service.
+A special section is dedicated at the end regarding databases.
+
+TODO: section on RuntimeDir, StateDir, WorkingDir
+
+### Database connection {#sec-systemd-state-management-database-connection}
+
+Most databases have 2 modes of connection, over TCP or over Unix socket. Connecting over a unix socket can provide 2 advantages
+- 30% performance improvement as packets don't need to be encoder over TCP
+- The default authentication mode postgres on unix sockets is just with the user name. This will remove the need to manage the password secret
```
Comment (roberth) [inline nixos/doc/manual/development/systemd-state-management.section.md:12]:
```suggestion
- The default authentication mode postgres on unix sockets relies on the client's uid provided by the kernel, rather than a password that you need to be care for. 
```

Comment (roberth) [reply nixos/doc/manual/development/systemd-state-management.section.md:12]: Though it being based on uid, this may be tricky across uid namespaces and therefore containers. I vaguely recall a nixpkgs issue or stackoverflow question about this, but I couldn't find it.

File: nixos/doc/manual/development/systemd.chapter.md [hunk 4]
```diff
@@ -0,0 +1,51 @@
+# Systemd  {#sec-systemd}
+
+Systemd is a building block of most nixos modules.
+
+From the [project's webpage](https://systemd.io/)
+```
```
Comment (roberth) [inline nixos/doc/manual/development/systemd.chapter.md:6]: This isn't code but a quote. You should probably use `> ` syntax instead.

**Comment by colemickens** [comment]
A react emoji isn't enough -- thank you for this. this will be invaluable to have.

File: nixos/doc/manual/development/systemd-handling-dependencies.section.md [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+
+PartOf:
+- use `partOf = [ "B" ];`, if service A needs to re-started when service B is.
+
+Note that `wants,requires,bindsTo` are orthogonal to `after`. The most common occurence in nixpkgs is `requires` and `after`.
+
+### Examples {#sec-systemd-handling-dependencies-examples}
+
+- Web-app dependency on a database. If the web-app cannot start without the database, use `requires` and `after`. If you would rather have it operate even without a database use `wants` and `after`. [example in nixpkgs](https://github.com/NixOS/nixpkgs/blob/master/nixos/modules/services/web-apps/lemmy.nix#L169)
+- Database migration and setup script for a web-app. It is common to make a service for database setup and migration. This service should be `partOf` the webapp service as everytime the webapp is restarted, this service should be restarted as well.
```
Comment (evils) [inline nixos/doc/manual/development/systemd-handling-dependencies.section.md:21]:
```suggestion
- Database migration and setup script for a web-app. It is common to make a service for database setup and migration. This service should be `partOf` the web-app service as every time the web-app restarts, this service should be restarted as well.
```

File: nixos/doc/manual/development/systemd-handling-dependencies.section.md [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+[example in nixpkgs](https://github.com/NixOS/nixpkgs/blob/master/nixos/modules/services/web-apps/lemmy.nix#L214)
+- A web-app has two separate services, a UI and a server. The ui `requires` `after` the server. [example in nixpkgs](https://github.com/NixOS/nixpkgs/blob/master/nixos/modules/services/web-apps/lemmy.nix#L205)
```
Comment (evils) [inline nixos/doc/manual/development/systemd-handling-dependencies.section.md:23]: this is hard to parse as a sentence, also hard to keep it short :)
```suggestion
- A web-app with two separate services, a UI and a server. The UI `requires` the server and starts `after` it. [example in nixpkgs](https://github.com/NixOS/nixpkgs/blob/master/nixos/modules/services/web-apps/lemmy.nix#L205)
```

File: nixos/doc/manual/development/systemd-state-management.section.md [hunk 7]
```diff
@@ -0,0 +1,24 @@
...
+
+Therefore the default for database connection for services is recommended to be with a unix socket. For postgres for example, the `host` needs to start with a `/`.
```
Comment (evils) [inline nixos/doc/manual/development/systemd-state-management.section.md:14]:
```suggestion
Therefore the recommended default for database connections for services is a unix socket. With postgres for example, `host` needs to start with a `/`.
```

File: nixos/doc/manual/development/systemd-state-management.section.md [hunk 8]
```diff
@@ -0,0 +1,24 @@
...
+[nixpkgs-example](https://github.com/NixOS/nixpkgs/blob/master/nixos/modules/services/web-apps/lemmy.nix#L98)
+
+### Database initialisation {#sec-systemd-state-management-database-initialisation}
+
+To initialise a database that a service depends on, it's customary to create another systemd service that will be `partOf` your main service.
+The initialisation service should be run by the database superuser so as not to require sudo.
+Typically it will involve creating the database and the user.
+There will be a step to see if the database is already created.
+That check is best implemented by going throug the created tables and exiting if the table already exists.
```
Comment (evils) [inline nixos/doc/manual/development/systemd-state-management.section.md:23]:
```suggestion
That check is best implemented by going through the created tables and exiting if the table already exists.
```

**Review by evils — COMMENTED** [review]
some style points that caused me to hesitate while reading

it may be a bit early to start nitpicking
please ignore all of this if it'll take away time from substantive work
(especially the spell checking, which is non-exhaustive)
i'd love to do a more in-depth pass when this is closer to completion (or after it's merged)

thanks for doing this, i look forward to having it in the manual

**Comment by con-f-use** [comment]
Any reason this hasn't been merged?

**Comment by happysalada** [comment]
This was too early in the process its just incomplete. If anyone ever wants to take it up, they are more than welcome to take any part they want

**Comment by ibizaman** [comment]
FYI I'm working on https://github.com/NixOS/nixpkgs/pull/328472 and will pick this PR up afterwards if no one beats me to it, at least for the credentials part.

EDIT: Famous last words :sob:

File: nixos/doc/manual/from_md/development/systemd-environment-and-secrets.section.xml [hunk 9]
```diff
@@ -0,0 +1,61 @@
+<section xmlns="http://docbook.org/ns/docbook" xmlns:xlink="http://www.w3.org/1999/xlink" xml:id="sec-systemd-environment">
+  <title>Environment and secret maangement</title>
```
Comment (cjshearer) [inline nixos/doc/manual/from_md/development/systemd-environment-and-secrets.section.xml:2]: maangement -> management
